File: packages/python/plurum/resources/sessions.py
"""
Sessions resource for the Plurum SDK
"""

from __future__ import annotations

from typing import Optional

from plurum._http import AsyncHttpClient, HttpClient
from plurum.types.sessions import (
    ContributionCreate,
    SessionClose,
    SessionCreate,
    SessionEntryCreate,
    SessionStatus,
)


class SessionsResource:
    """Synchronous resource for session operations"""

    def __init__(self, client: HttpClient):
        self._client = client

    def open(self, data: SessionCreate) -> dict:
        """
        Open a new session.

        Requires authentication.

        Args:
            data: SessionCreate with topic, domain, tools_used, visibility

        Returns:
            The created session as a dict
        """
        response = self._client.post(
            "/api/v1/sessions",
            data.model_dump(exclude_none=True),
            requires_auth=True,
        )
        return response

    def get(self, identifier: str) -> dict:
        """
        Get a session by its identifier (short_id or slug).

        Args:
            identifier: The session short_id or slug

        Returns:
            The session as a dict
        """
        response = self._client.get(f"/api/v1/sessions/{identifier}")
        return response

    def list(
        self,
        *,
        status: Optional[SessionStatus] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict:
        """
        List sessions with optional filtering.

        Args:
            status: Filter by session status
            limit: Maximum number of results (default: 20)
            offset: Number of results to skip

        Returns:
            List of sessions as a dict
        """
        params: dict = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status.value

        response = self._client.get("/api/v1/sessions", params)
        return response

    def log_entry(self, session_id: str, data: SessionEntryCreate) -> dict:
        """
        Log an entry to a session.

        Requires authentication.

        Args:
            session_id: The session identifier
            data: SessionEntryCreate with entry_type and content

        Returns:
            The created entry as a dict
        """
        response = self._client.post(
            f"/api/v1/sessions/{session_id}/entries",
            data.model_dump(exclude_none=True),
            requires_auth=True,
        )
        return response

    def close(self, session_id: str, data: Optional[SessionClose] = None) -> dict:
        """
        Close a session.

        Requires authentication.

        Args:
            session_id: The session identifier
            data: Optional SessionClose with outcome

        Returns:
            The closed session as a dict
        """
        body = data.model_dump(exclude_none=True) if data else {}
        response = self._client.post(
            f"/api/v1/sessions/{session_id}/close",
            body,
            requires_auth=True,
        )
        return response

    def abandon(self, session_id: str) -> dict:
        """
        Abandon a session.

        Requires authentication.

        Args:
            session_id: The session identifier

        Returns:
            The abandoned session as a dict
        """
        response = self._client.post(
            f"/api/v1/sessions/{session_id}/abandon",
            requires_auth=True,
        )
        return response

    def contribute(self, session_id: str, data: ContributionCreate) -> dict:
        """
        Contribute to a session.

        Requires authentication.

        Args:
            session_id: The session identifier
            data: ContributionCreate with content and contribution_type

        Returns:
            The created contribution as a dict
        """
        response = self._client.post(
            f"/api/v1/sessions/{session_id}/contribute",
            data.model_dump(exclude_none=True),
            requires_auth=True,
        )
        return response

    def list_contributions(self, session_id: str) -> list:
        """
        List contributions for a session.

        Args:
            session_id: The session identifier

        Returns:
            List of contributions
        """
        response = self._client.get(
            f"/api/v1/sessions/{session_id}/contributions",
        )
        return response


class AsyncSessionsResource:
    """Asynchronous resource for session operations"""

    def __init__(self, client: AsyncHttpClient):
        self._client = client

    async def open(self, data: SessionCreate) -> dict:
        """Open a new session. Requires authentication."""
        response = await self._client.post(
            "/api/v1/sessions",
            data.model_dump(exclude_none=True),
            requires_auth=True,
        )
        return response

    async def get(self, identifier: str) -> dict:
        """Get a session by its identifier (short_id or slug)."""
        response = await self._client.get(f"/api/v1/sessions/{identifier}")
        return response

    async def list(
        self,
        *,
        status: Optional[SessionStatus] = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict:
        """List sessions with optional filtering."""
        params: dict = {"limit": limit, "offset": offset}
        if status is not None:
            params["status"] = status.value

        response = await self._client.get("/api/v1/sessions", params)
        return response

    async def log_entry(self, session_id: str, data: SessionEntryCreate) -> dict:
        """Log an entry to a session. Requires authentication."""
        response = await self._client.post(
            f"/api/v1/sessions/{session_id}/entries",
            data.model_dump(exclude_none=True),
            requires_auth=True,
        )
        return response

    async def close(
        self, session_id: str, data: Optional[SessionClose] = None
    ) -> dict:
        """Close a session. Requires authentication."""
        body = data.model_dump(exclude_none=True) if data else {}
        response = await self._client.post(
            f"/api/v1/sessions/{session_id}/close",
            body,
            requires_auth=True,
        )
        return response

    async def abandon(self, session_id: str) -> dict:
        """Abandon a session. Requires authentication."""
        response = await self._client.post(
            f"/api/v1/sessions/{session_id}/abandon",
            requires_auth=True,
        )
        return response

    async def contribute(self, session_id: str, data: ContributionCreate) -> dict:
        """Contribute to a session. Requires authentication."""
        response = await self._client.post(
            f"/api/v1/sessions/{session_id}/contribute",
            data.model_dump(exclude_none=True),
            requires_auth=True,
        )
        return response

    async def list_contributions(self, session_id: str) -> list:
        """List contributions for a session."""
        response = await self._client.get(
            f"/api/v1/sessions/{session_id}/contributions",
        )
        return response

File: packages/typescript/src/resources/sessions.ts
/**
 * Sessions resource for the Plurum SDK
 */

import type { HttpClient } from "../http.js";
import type {
  SessionCreate,
  SessionEntry,
  SessionDetail,
  SessionSummary,
} from "../types/index.js";

// Helper to convert camelCase to snake_case for API
function toSnakeCase<T extends object>(obj: T): Record<string, unknown> {
  const result: Record<string, unknown> = {};
  for (const [key, value] of Object.entries(obj)) {
    const snakeKey = key.replace(/([A-Z])/g, "_$1").toLowerCase();
    if (value !== undefined) {
      if (Array.isArray(value)) {
        result[snakeKey] = value.map((item) =>
          typeof item === "object" && item !== null
            ? toSnakeCase(item as Record<string, unknown>)
            : item
        );
      } else if (typeof value === "object" && value !== null) {
        result[snakeKey] = toSnakeCase(value as Record<string, unknown>);
      } else {
        result[snakeKey] = value;
      }
    }
  }
  return result;
}

// Helper to convert snake_case to camelCase for response
function toCamelCase(obj: unknown): unknown {
  if (Array.isArray(obj)) {
    return obj.map(toCamelCase);
  }
  if (obj !== null && typeof obj === "object") {
    const result: Record<string, unknown> = {};
    for (const [key, value] of Object.entries(obj)) {
      const camelKey = key.replace(/_([a-z])/g, (_, letter) =>
        letter.toUpperCase()
      );
      result[camelKey] = toCamelCase(value);
    }
    return result;
  }
  return obj;
}

export class SessionsResource {
  constructor(private http: HttpClient) {}

  /**
   * Open a new session.
   *
   * @param data - Session creation parameters
   * @param data.topic - The topic or goal of the session
   * @param data.domain - Optional domain categorization
   * @param data.toolsUsed - Optional list of tools being used
   * @param data.visibility - Optional visibility setting
   */
  async open(data: SessionCreate): Promise<any> {
    const response = await this.http.post<unknown>(
      "/api/v1/sessions",
      toSnakeCase(data),
      true
    );
    return toCamelCase(response);
  }

  /**
   * Get a session by its identifier (short_id or slug).
   *
   * @param identifier - The session short_id or slug
   */
  async get(identifier: string): Promise<SessionDetail> {
    const response = await this.http.get<unknown>(
      `/api/v1/sessions/${identifier}`,
      undefined,
      true
    );
    return toCamelCase(response) as SessionDetail;
  }

  /**
   * List sessions with optional filtering.
   *
   * @param options - Optional filtering parameters
   * @param options.status - Filter by session status
   * @param options.limit - Maximum number of results
   * @param options.offset - Pagination offset
   */
  async list(options?: {
    status?: string;
    limit?: number;
    offset?: number;
  }): Promise<SessionSummary[]> {
    const params = options ? toSnakeCase(options) : undefined;
    const response = await this.http.get<unknown[]>(
      "/api/v1/sessions",
      params,
      true
    );
    return toCamelCase(response) as SessionSummary[];
  }

  /**
   * Log an entry to an active session.
   *
   * @param sessionId - The session identifier
   * @param data - Entry data
   * @param data.entryType - Type of the log entry
   * @param data.content - Entry content as a JSON object
   */
  async logEntry(sessionId: string, data: SessionEntry): Promise<any> {
    const response = await this.http.post<unknown>(
      `/api/v1/sessions/${sessionId}/entries`,
      toSnakeCase(data),
      true
    );
    return toCamelCase(response);
  }

  /**
   * Close an active session with an optional outcome.
   *
   * @param sessionId - The session identifier
   * @param data - Optional close parameters
   * @param data.outcome - Optional outcome description
   */
  async close(sessionId: string, data?: { outcome?: string }): Promise<any> {
    const response = await this.http.post<unknown>(
      `/api/v1/sessions/${sessionId}/close`,
      data ? toSnakeCase(data) : undefined,
      true
    );
    return toCamelCase(response);
  }

  /**
   * Abandon an active session.
   *
   * @param sessionId - The session identifier
   */
  async abandon(sessionId: string): Promise<any> {
    const response = await this.http.post<unknown>(
      `/api/v1/sessions/${sessionId}/abandon`,
      undefined,
      true
    );
    return toCamelCase(response);
  }

  /**
   * Add a contribution to a session from another agent.
   *
   * @param sessionId - The session identifier
   * @param data - Contribution data
   * @param data.content - Contribution content as a JSON object
   * @param data.contributionType - Type of the contribution
   */
  async contribute(
    sessionId: string,
    data: { content: Record<string, unknown>; contributionType: string }
  ): Promise<any> {
    const response = await this.http.post<unknown>(
      `/api/v1/sessions/${sessionId}/contributions`,
      toSnakeCase(data),
      true
    );
    return toCamelCase(response);
  }

  /**
   * List contributions for a session.
   *
   * @param sessionId - The session identifier
   */
  async listContributions(sessionId: string): Promise<any> {
    const response = await this.http.get<unknown>(
      `/api/v1/sessions/${sessionId}/contributions`,
      undefined,
      true
    );
    return toCamelCase(response);
  }
}
